src/sentry/snuba/entity_subscription.py

```diff
@@ -198,20 +204,36 @@ class MetricsCountersEntitySubscription(BaseEntitySubscription):
     dataset = Dataset.Metrics.value
     entity_key = EntityKey.MetricsCounters.value
 
-    def __init__(self, aggregate: str, extra_fields: Optional[Mapping[str, Any]] = None):
-        super().__init__(aggregate, extra_fields)
+    def __init__(
+        self, aggregate: str, time_window: int, extra_fields: Optional[Mapping[str, Any]] = None
+    ):
+        super().__init__(aggregate, time_window, extra_fields)
         self.aggregate = aggregate
         if not extra_fields or "org_id" not in extra_fields:
             raise InvalidQuerySubscription(
                 "org_id is a required param when "
                 "building snuba filter for a metrics subscription"
             )
         self.org_id = extra_fields["org_id"]
+        self.time_window = time_window
 
     def get_query_groupby(self) -> List[str]:
         session_status = tag_key(self.org_id, "session.status")
         return ["project_id", session_status]
 
+    def get_granularity(self):
+        # Both time_window and granularity are in seconds
+        # Time windows <= 1h -> Granularity 10s
+        # Time windows > 1h & <= 4h -> Granularity 60s
+        # Time windows > 4h -> Granularity 1 hour
+        if self.time_window <= 3600:
+            granularity = 10
+        elif self.time_window <= 4 * 3600:
+            granularity = 60
+        else:
+            granularity = 3600
+        return granularity
```

Comment:
Metrics also have daily aggregation, would it make sense to set `granularity = 24 * 60 * 60` for very large time windows?

Reply:
See comment above